docs/latest/design/extending-envoy-gateway.md

```diff
@@ -0,0 +1,176 @@
+An extension server can make use of one or more pre/post hooks inside Envoy Gateway before and after it's major components (translator, etc.) to allow the extension to modify the data going into or coming out of these components.
+An extension can be created external to Envoy Gateway as it's own Kubernetes deployment or loaded as a sidecar. gRPC is used for the calls between Envoy Gateway and an extension. In the hook call, Envoy Gateway sends data as well 
+as context information to the extension and expects a reply with a modified version of the data that was sent to the extension. Since extensions fundamentally alter the logic and data that Envoy Gateway provides, Extension projects assume responsibility for any bugs and issues 
+they create as a direct result of their modification of Envoy Gateway.
+
+## Diagram
+
+![Architecture](../images/extension.png)
+
+## Registering Extensions in Envoy Gateway
+
+Information about the extensions that Envoy Gateway needs to load is configured in the Envoy Gateway config. Each extension must specify:
+
+- The name to register the extension under
+- Which API groups the extension supports
+- The `Kind`s supported under each API group
+- Extension service connection info (host, port, TLS settings, etc.)
+
+An example configuration:
+
+```yaml
+apiVersion: config.gateway.envoyproxy.io/v1alpha1
+kind: EnvoyGateway
+extensions:
+  - name: example-extension
+    apiGroups:
+      - apiGroup: example.myextension.io
+        kinds:
+          - OAuth2Filter
+    service:
+      host: my-extension.example
+      port: 443
+      tls:
+        type: secret
+        secret:
+            name: my-extension
+            namespace: example
+```
+
+This configuration is required to be provided at bootstrap and modifying the registered extensions during runtime is not currently supported.
+Envoy Gateway will keep track of the registered extensions and their `apiGroups` and `kinds` when processing Gateway API resources.
+
+## Extending Gateway API and the Data Plane
+
+Envoy Gateway manages [Envoy][] deployments, which act as the data plane that handles actual user traffic. Users configure the data plane using the K8s Gateway API resources which Envoy 
+Gateway converts into [Envoy specific configuration (xDS)][] to send over to Envoy.
+
+Gateway API offers [ExtensionRef filters][] and [Policy Attachments][] as extension points for implementers to use. Envoy Gateway extends the Gateway API using these extension points to provide support for [rate limiting][] 
+and [authentication][]. The initial design of Envoy Gateway extensions will primarily focus on ExtensionRef filters, where users can reference a custom resource as an HTTP Filter.
+
+When Envoy Gateway encounters an [HTTPRoute][] or [GRPCRoute][] that has an `ExtensionRef` `filter` with an `apiGroup` and `kind` that Envoy Gateway does not supoport, it will first
```

Comment:
```suggestion
When Envoy Gateway encounters an [HTTPRoute][] or [GRPCRoute][] that has an `ExtensionRef` `filter` with an `apiGroup` and `kind` that Envoy Gateway does not support, it will first
```